Memory scheduling deadlock

jaewan opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

## Why are these changes needed?

<!-- Please give a short summary of the change and the problem this solves. -->
Implementation of Deadlock Detection1.
This checks the number of leased_workers and spinning_tasks(blocked tasks from triggering spill)
When the numbers match, consider it as a deadlock and trigger spill.

jaewan committed:
blocktask and evict tasks separated

jaewan committed:
blocktasks when oom

jaewan committed:
command to replay the bug

stephanie-wang committed:
Fix crash in raylet on duplicate object

jaewan committed:
Fixed bug in blocktasks. Did not block when block task called only

jaewan committed:
Merged blocktasks and evicttasks into a single callback and erased evict callback

jaewan committed:
get lowest pri, but obj does not have pri yet

jaewan committed:
debug logs to see where to put priority

Commit:
log without compile error

jaewan committed:
commit for migration

jaewan committed:
working version. But still need testing

jaewan committed:
Blocktasks but 2 new problems emerged

jaewan committed:
added microbechmark

jaewan committed:
deadlock cases 1 and 2 induced by not spilling. 1 is when workers are stuck with producers and 2 is when consumers are dependent on multiple objects

jaewan committed:
polished debug messages

jaewan committed:
typo fix

jaewan committed:
removed scripts and included block, evict threshold triggered code

jaewan committed:
deadlock detection by checking #of leased workers and spinning tasks

jaewan committed:
reflected changes from branch -jae

jaewan committed:
reflected changes from branch -jae

jaewan committed:
deadlock detection#1

jaewan committed:
reflected changes from parent branch

jaewan committed:
.

jaewan committed:
Merge branch 'memory-scheduling' into memory-scheduling-deadlock

jaewan committed:
Merge branch 'memory-scheduling' into memory-scheduling-deadlock

stephanie-wang reviewed (changes requested):
Two things:
1. Ideally, we shouldn't leak the number of spilled workers to the plasma request queue. Instead, can we do this logic in the task manager? The reason is that we already has a way for the task manager to notify the request queue when to block/allow spill, so we can build off of this codepath instead of creating a new one.
2. Can you add a test in Python that demonstrates the new behavior? We want one that previously did not work when blocking spills was enabled, and now it does work. Also, we should compare the runtime to the default Ray implementation that always spills.

jaewan committed:
changed triggering spill from task_manager instead of giving #leased_worker to the obj store

jaewan commented:
> Two things:
> 
> 1. Ideally, we shouldn't leak the number of spilled workers to the plasma request queue. Instead, can we do this logic in the task manager? The reason is that we already has a way for the task manager to notify the request queue when to block/allow spill, so we can build off of this codepath instead of creating a new one.
> 2. Can you add a test in Python that demonstrates the new behavior? We want one that previously did not work when blocking spills was enabled, and now it does work. Also, we should compare the runtime to the default Ray implementation that always spills.

1. I changed the implementation to trigger spill from task_manager. This is done by setting should_spill_ in object store via io_post. 
    !!! Can you check this Stephanie? is this design OK or should I get back to previous design?
    1.1 I also changed the setting should_spill_ from task_manager happens only when it is true (1. When #spinning_workers==#leased_workers, 2. When evict_tasks() called but did not evict anything). In previous design should_spill_ is always set after calling evict_tasks() and block_spill(). So no need for resetting should_spill_ from the object store. Advantage of this is since the calls are asynchronous, there's a time difference between setting should_spill_ from the task_manager to object_store. More spill() will be trigger in the meantime. But I thought it's wasting communication cost so changed the implementation. When should_spill_ is set, object_manager triggers spill once and set it off so task_manager doesn't need to reset it. 

2. We have a separate PR for test scripts. Do you want me to add test of this PR here or to the test script PR?

jaewan committed:
Spinning task identificatino by queue_ version.

jaewan committed:
Merge branch 'memory-scheduling' into memory-scheduling-deadlock

stephanie-wang commented on src/ray/object_manager/plasma/create_request_queue.cc:
```diff
@@ -196,19 +202,32 @@ Status CreateRequestQueue::ProcessRequests() {
-	      on_object_creation_blocked_callback_(lowest_pri, enable_blocktasks , enable_evicttasks);
+		if(enable_blocktasks_spill){
+          RAY_LOG(DEBUG) << "[JAE_DEBUG] task " << task_id << " spins"; 
+		  for(auto it = queue_.begin(); it != queue_.end(); it++){
```
Can't we just use queue.size()?

stephanie-wang commented:
The implementation looks much better!

> We have a separate PR for test scripts. Do you want me to add test of this PR here or to the test script PR?

Can you add just one script that previously deadlocked and now works with this PR?

Commit:
test rpc to core worker for object working set

jaewan committed:
Revert "test rpc to core worker for object working set"

This reverts commit 72e3543a293104b68a41c55b0cfc4e92dd74b4cc.

jaewan committed:
test caese, will be updated to pytest
